Prepare release 0.41.0 (#2340)
Qiskit/qiskit-ibm-runtime: release-notes/0.41.0.rst

0.41.0 (2025-07-24)
===================

Upgrade Notes
-------------

- Because of the sunset of IBM Quantum Platform Classic, the ``ibm_quantum`` channel is no
  longer supported from ``qiskit-ibm-runtime``. Saved ``ibm_quantum`` channel accounts and 
  data will not be accessible. Use the ``ibm_quantum_platform`` channel instead. See our 
  [migration guide](https://docs.quantum.ibm.com/migration-guides/classic-iqp-to-cloud-iqp) 
  for more details. (`2289 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2289>`__)

New Features
------------

> - A new parameter ``tags``, has been added to the 

Reviewer: ```suggestion
- A new parameter, ``tags``, has been added to the 
```

  :class:`.QiskitRuntimeService` initializer and :meth:`.QiskitRuntimeService.save_account` method. 
  ``tags`` can be used to **filter** instances so only instances with the given tags are 
  returned.

>   Additionally, if using the filters  ``tags``, ``region``, or ``plans_preference`` returns no valid instances, 

Reviewer 2: ```suggestion
  Additionally, if no valid instances are returned when using the ``tags``, ``region``, or ``plans_preference`` filters, 
```

>   an error will now be raised at initialization. Make sure the name of the ``tags`` and ``region`` passed in are 

Reviewer 2: ```suggestion
  an error will now be raised at initialization. Make sure the names of the ``tags`` and ``region`` passed in 
```

>   an exact match of the instance ``tags`` or ``region`` (case insensitive). For ``plans_preference``, as long as a matching

Reviewer 2: ```suggestion
  match the instance ``tags`` or ``region`` exactly (case insensitive). For ``plans_preference``, as long as a matching
```

  plan name is passed in, instances with the matching plan name will be prioritized. (`2277 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2277>`__)

> - With the migration to the new IBM Quantum Platform, there were a few inconsistencies with the old IBM Quantum 

Reviewer 2: ```suggestion
- With the migration to the new IBM Quantum Platform, there were a few inconsistencies with IBM Quantum 
```

>   Platform classic that needed to be addressed. 

Reviewer 2: ```suggestion
  Platform Classic that needed to be addressed:
```

      - :meth:`.QiskitRuntimeService.usage` now returns usage information for the current active IBM Cloud instance.

>       - :meth:`.QiskitRuntimeService.delete_job` is not supported on the new platform so it has been deprecated.

Reviewer 2: ```suggestion
      - :meth:`.QiskitRuntimeService.delete_job` is not supported on the new platform, so it has been deprecated.
```

>       - :meth:`.RuntimeJobV2.instance` will now return the IBM Cloud instance crn. 

Reviewer 2: ```suggestion
      - :meth:`.RuntimeJobV2.instance` will now return the IBM Cloud instance CRN. 
```

>       - :meth:`.RuntimeJob.queue_usage` is also not supported on the new platform so it has been deprecated. (`2296 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2296>`__)

Reviewer 2: ```suggestion
      - :meth:`.RuntimeJob.queue_usage` is not supported on the new platform, so it has been deprecated. (`2296 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2296>`__)
```

- Support for running ``qiskit-ibm-runtime`` with Python 3.9 has been deprecated and will
  be removed in a future release. 

  Additionally, ``qiskit-ibm-runtime`` now officialy supports Python 3.13. (`2314 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2314>`__)
- When saving an account, there is now a validation check to make sure the ``region``, 

>   ``plans_preference`` and ``tags`` parameters are valid. (`2319 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2319>`__)

Reviewer 2: ```suggestion
  ``plans_preference``, and ``tags`` parameters are valid. (`2319 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2319>`__)
```

- There is a new method, :meth:`.QiskitRuntimeService.active_instance` which returns the IBM Cloud CRN 
  of the current active instance. 

>   The :meth:`.QiskitRuntimeService.jobs` method has also been updated to properly accept the ``instance`` 

Reviewer 2: ```suggestion
  The :meth:`.QiskitRuntimeService.jobs` method has been updated to properly accept the ``instance`` 
```

  parameter, which can be used to filter jobs. (`2325 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2325>`__)
- :meth:`.QiskitRuntimeService.usage` has been updated to include a ``usage_remaining_seconds`` field. 

>   This field includes the difference between the reported ``usage_limit_seconds`` / ``usage_allocation_seconds`` 

Reviewer 2: ```suggestion
  This field includes the difference between the reported ``usage_limit_seconds`` or  ``usage_allocation_seconds`` 
```

  (depending on how the instance is configured) and ``usage_consumed_seconds``. 

>   Learn more about instance allocation limits `here <https://quantum.cloud.ibm.com/docs/en/guides/allocation-limits>`__. (`2329 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2329>`__)

Reviewer 2: ```suggestion
  Learn more about instance allocation limits `here <https://quantum.cloud.ibm.com/docs/guides/allocation-limits>`__. (`2329 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2329>`__)
```

Bug Fixes
---------

- Fixed a bug in :class:`.BasePrimitive` where primitives instantiated inside a :class:`~.Session` or :class:`~.Batch` context manager without the ``mode`` 
  parameter would fetch the backend from the service (by name) instead of using the backend passed to the :class:`~.Session` or :class:`~.Batch`. 

>   This could cause issues when the :class:`~.Session`/:class:`~.Batch`

Reviewer 2: ```suggestion
  This could cause issues when the :class:`~.Session` or :class:`~.Batch`
```

  backend was modified by users (for example, by removing a gate), because the primitives 
  would instead fetch the unmodified backend object from the service. After the fix, the

>   :class:`~.Session`/:class:`~.Batch` backend object is used directly. (`2282 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2282>`__)

Reviewer 2: ```suggestion
  :class:`~.Session` or :class:`~.Batch` backend object is used directly. (`2282 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2282>`__)
```

- Fixed an issue in :class:`.IBMBackend` where property changes, such as modifications to ``basis_gates``, persisted even after the backend object was renewed. (`2283 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2283>`__)

> Other Notes

Reviewer 2: ```suggestion
Other notes
```

-----------

> - The deprecated :class:`.IBMBackend` attributes ``max_shots`` and 

Reviewer: ```suggestion
- The deprecated :class:`.IBMBackend` attributes, ``max_shots`` and 
```

  ``max_experiments``, have been removed and the :meth:`.IBMBackend.max_circuits` method now
  returns ``None``. See the `job limits guide <https://quantum.cloud.ibm.com/docs/guides/job-limits#job-limits>`__ for details. (`2235 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2235>`__)
- The deprecated :class:`~.IBMBackend` and :class:`~.FakeBackendV2` ``defaults()`` methods 
  have been removed. They were deprecated in the 0.38.0 release. 
  Pulse defaults have also been removed from all fake backends. (`2238 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2238>`__)

> - Improved error message when :meth:`.ExecutionSpan.mask` is called for a pub that's not contained in the span. (`2311 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2311>`__)

Reviewer 2: ```suggestion
- Improved the error message returned when :meth:`.ExecutionSpan.mask` is called for a PUB that's not contained in the span. (`2311 <https://github.com/Qiskit/qiskit-ibm-runtime/pull/2311>`__)
```
